### PR TITLE
HttpRequest functional, but the return has a JSONexception

### PR DESCRIPTION
**Pull request by Oulala-Leon**
First Pull Request, just some cleaning.

JSONException is thrown:

org.json.JSONException: Value [{"id":1990054,"valid":true,"title":"Dossier Brevet","url":"https:\/\/lls-media-public.s3-eu-west-1.amazonaws.com\/svt-3e-2017\/dossier-brevet\/1200.llsa-asset-acces-kit-bac.jpg","number":0,"numShow":"0","book":1339497,"parent":1339497,"lessons":[2005907,2006188,2006154,2006086,2006052,2005952,2005951,2005944,2005942,2006120,2006017,2005949]},{"id":1339669,"valid":true,"title":"Le contrôle de la reproduction","url":"https:\/\/lls-media-public.s3-eu-west-1.amazonaws.com\/svt3e-2017\/le-controle-de-la-reproduction\/svt.t3.31.ouv.fecondation-vitro-01.png","colorTheme":"corps","teacherBook":"https:\/\/lls-media-public.s3-eu-west-1.amazonaws.com\/svt-3e-2017\/le-controle-de-la-reproduction\/livre-prof-svtc4-chap31.pdf","themeId":1339507,"theme":"3 - Le corps humain et la santé","number":11,"numShow":"11","book":1339497,"parent":1339497,"lessons":[1342288,1342286,1342205,1342212,1342239,1342268,1342235,1342177,1342204]},{"id":1339665,"valid":true,"title":"La réponse immunitaire adaptative face aux infections","url":"https:\/\/lls-media-public.s3-eu-west-1.amazonaws.com\/svt3e-2017\/la-reponse-immunitaire-adaptative-face-aux-infections\/svt.t3.28.ouv.vaccination.jpg","colorTheme":"corps","teacherBook":"https:\/\/lls-media-public.s3-eu-west-1.amazonaws.com\/svt-3e-2017\/la-reponse-immunitaire-adaptative-face-aux-infections\/livre-prof-svtc4-chap28.pdf","themeId":1339507,"theme":"3 - Le corps humain et la santé","number":10,"numShow":"10","book":1339497,"parent":1339497,"lessons":[1341648,1341882,1342060,1342064,1341963,1341671,1341647]},{"id":1339659,"valid":true,"title":"Des aliments aux nutriments","url":"https:\/\/lls-media-public.s3-eu-west-1.amazonaws.com\/svt3e-2017\/des-aliments-aux-nutriments\/svt.t3.25.ouv.miyazaki.jpg","colorTheme":"corps","teacherBook":"https:\/\/lls-media-public.s3-eu-west-1.amazonaws.com\/svt-3e-2017\/des-aliments-aux-nutriments\/livre-prof-svtc4-chap25.pdf","themeId":1339507,"theme":"3 - Le corps humain et la santé","number":9,"numShow":"9","book":1339497,"parent":1339497,"lessons":[1341560,1341559,1341557,1341555,1341558,1341518,1341517]},{"id":1339658,"valid":true,"title":"Le fonctionnement du système nerveux","url":"https:\/\/lls-media-public.s3-eu-west-1.amazonaws.com\/svt3e-2017\/le-fonctionnement-du-systeme-nerveux\/svt.t3.22.ouv.neuronne.jpg","colorTheme":"corps","teacherBook":"https:\/\/lls-media-public.s3-eu-west-1.amazonaws.com\/svt-3e-2017\/le-fonctionnement-du-systeme-nerveux\/livre-prof-svtc4-chap22.pdf","themeId":1339507,"theme":"3 - Le corps humain et la santé","number":8,"numShow":"8","book":1339497,"parent":1339497,"lessons":[1341474,1341476,1341481,1341514,1341478,1341473]},{"id":1339647,"valid":true,"title":"L'évolution de la biodiversité","url":"https:\/\/lls-media-public.s3-eu-west-1.amazonaws.com\/svt3e-2017\/l-evolution-de-la-biodiversite\/svt.t2.20.ouv.colibris.jpg","colorTheme":"vivant","teacherBook":"https:\/\/lls-media-public.s3-eu-west-1.amazonaws.com\/svt-3e-2017\/l-evolution-de-la-biodiversite\/livre-prof-svtc4-chap20.pdf","themeId":1339499,"theme":"2 - Le vivant et son évolution","number":7,"numShow":"7","book":1339497,"parent":1339497,"lessons":[1341421,1341420,1341403,1341401,1341400,1341410,1341402,1341398,1341399]},{"id":1339638,"valid":true,"title":"De la diversité génétique à la biodiversité","url":"https:\/\/lls-media-public.s3-eu-west-1.amazonaws.com\/svt3e-2017\/de-la-diversite-genetique-a-la-biodiversite\/svt.t2.17.ouv.chevaux.jpg","colorTheme":"vivant","teacherBook":"https:\/\/lls-media-public.s3-eu-west-1.amazonaws.com\/svt-3e-2017\/de-la-diversite-genetique-a-la-biodiversite\/livre-prof-svtc4-chap17.pdf","themeId":1339499,"theme":"2 - Le vivant et son évolution","number":6,"numShow":"6","book":1339497,"parent":1339497,"lessons":[1341084,1341090,1341195,1341396,1341148,1341077,1341083]},{"id":1339594,"valid":true,"title":"La reproduction et le peuplement des milieux","url":"https:\/\/lls-media-public.s3-eu-west-1.amazonaws.com\/svt3e-2017\/la-reproduction-et-le-peuplemen
                  at org.json.JSON.typeMismatch(JSON.java:111)
                  at org.json.JSONObject.<init>(JSONObject.java:163)

It seems the HttpResponse is truncated compared to what curl has been sending back.

Will check that then focus on filling the first list.

### CODE REVIEW

**Comment by paour** [comment]
I think you're mixing synchronous and async styles… `getHttpResponse` shouldn't return a JSONObject since when it returns the request is ongoing.

The comment about 500 characters may be a clue to explain the truncated JSON string. I'm not familiar with Volley so I don't know why it would truncate.

**Comment by Oulala-Leon** [comment]
Trying to gradually get Async to callback on response. 

JSONArray has been fixed a few commits ago, but I decided to go for a Singleton design on the httpRequest to reuse it for other aspects of the program (multiple images for the views along with JSONarrays).

**Comment by Oulala-Leon** [comment]
Caused by: java.lang.ClassNotFoundException: Didn't find class "android.support.constraint.LinearLayout" on path: DexPathList

File: app/src/main/java/sirup/edubook_api_test/JSONAdapter.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    @Override public long getItemId(int position) {
+        JSONObject jsonObject = getItem(position);
+        return jsonObject.optLong("id");
+    }
+
+    @Override
+    public View getView(int position, View convertView, ViewGroup parent) {
+        if (convertView == null) {
+            LayoutInflater inflater = LayoutInflater.from(_context);
+            convertView = inflater.inflate(R.layout.chapterslist_row, parent); //possibly second argument should be null
```
[the remark this answers is not in the document]
Comment (Oulala-Leon) [reply app/src/main/java/sirup/edubook_api_test/JSONAdapter.java]: Fixed. Read the article.

File: app/src/main/java/sirup/edubook_api_test/JSONAdapter.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import android.widget.ListAdapter;
+import android.widget.TextView;
+
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+public class JSONAdapter extends BaseAdapter implements ListAdapter {
+
+    private JSONArray _jsonArray;
```
[the remark this answers is not in the document]
Comment (Oulala-Leon) [reply app/src/main/java/sirup/edubook_api_test/JSONAdapter.java]: Fixed. This convention was used in C++ for class variables. Will not use again.

File: app/src/main/java/sirup/edubook_api_test/JSONAdapter.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        if (convertView == null) {
+            LayoutInflater inflater = LayoutInflater.from(_context);
+            convertView = inflater.inflate(R.layout.chapterslist_row, parent); //possibly second argument should be null
+        }
+
+        TextView text = convertView.findViewById(R.id.chapter_title);
+        ImageView imageView = convertView.findViewById(R.id.chapter_image);
+
+        JSONObject json_data = getItem(position);
+        if(null!=json_data ){
```
[the remark this answers is not in the document]
Comment (Oulala-Leon) [reply app/src/main/java/sirup/edubook_api_test/JSONAdapter.java]: Ctrl Alt L properly learned. :)

No automatic reformatting options were found yet. Best agents on the case.

File: app/src/main/java/sirup/edubook_api_test/myHttpRequest.java [hunk 4]
```diff
@@ -1,72 +1,88 @@
...
+        }
+        return _instance;
+    }
 
-        // Request a string response from the provided URL.
-        StringRequest stringRequest = new StringRequest(Request.Method.GET, url,
-                new Response.Listener<String>() {
+    public static synchronized void setJSONArray(final String url, final ListView view) {
+
+        JsonArrayRequest JSONArrayRequest = new JsonArrayRequest(Request.Method.GET, url, null,
```
[the remark this answers is not in the document]
Comment (Oulala-Leon) [reply app/src/main/java/sirup/edubook_api_test/myHttpRequest.java]: Fixed.

File: app/src/main/res/layout/chapterslist_row.xml [hunk 5]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8"?>
+<android.support.constraint.LinearLayout
```
[the remark this answers is not in the document]
Comment (Oulala-Leon) [reply app/src/main/res/layout/chapterslist_row.xml]: Fixed.

**Comment by Oulala-Leon** [comment]
I am feeling a bit under the weather today, and have a big group meeting with my schoolmates at 19h.

I will fix/check out the rest tonight and tomorrow.

File: app/src/main/java/sirup/edubook_api_test/JSONAdapter.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+    @Override public long getItemId(int position) {
+        JSONObject jsonObject = getItem(position);
+        return jsonObject.optLong("id");
+    }
+
+    @Override
+    public View getView(int position, View convertView, ViewGroup parent) {
+        if (convertView == null) {
+            LayoutInflater inflater = LayoutInflater.from(_context);
```
[the remark this answers is not in the document]
Comment (Oulala-Leon) [reply app/src/main/java/sirup/edubook_api_test/JSONAdapter.java]: Looked up a UML of the Android Context architecture. Fixed.

File: app/src/main/java/sirup/edubook_api_test/JSONAdapter.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        TextView text = convertView.findViewById(R.id.chapter_title);
+        ImageView imageView = convertView.findViewById(R.id.chapter_image);
+
+        JSONObject json_data = getItem(position);
+        if(null!=json_data ){
+            try {
+                String title = json_data.getString("title");
+                text.setText(title);
+                String url = json_data.getString("url");
+                myHttpRequest.setImage(url, imageView);
```
[the remark this answers is not in the document]
Comment (Oulala-Leon) [reply app/src/main/java/sirup/edubook_api_test/JSONAdapter.java]: Modified architecture successfully. UI Code now calls the httpRequest instance.

File: app/src/main/java/sirup/edubook_api_test/myHttpRequest.java [hunk 8]
```diff
@@ -1,72 +1,88 @@
...
+            throw new IllegalStateException(myHttpRequest.class.getSimpleName() +
+            " is not initialised, call getInstance(context) first");
+        }
+        return _instance;
+    }
 
-        // Request a string response from the provided URL.
-        StringRequest stringRequest = new StringRequest(Request.Method.GET, url,
-                new Response.Listener<String>() {
+    public static synchronized void setJSONArray(final String url, final ListView view) {
```
[the remark this answers is not in the document]
Comment (Oulala-Leon) [reply app/src/main/java/sirup/edubook_api_test/myHttpRequest.java]: Done.

File: app/src/main/java/sirup/edubook_api_test/MainActivity.java [hunk 9]
```diff
@@ -2,43 +2,42 @@
...
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
 
-        String bookUrl = "https://api.lelivrescolaire.fr/public/books";
-        myHttpRequest request = new myHttpRequest();
+        //set Chapters
+        ListView listView = findViewById(R.id.Chapters_List);
```
[the remark this answers is not in the document]
Comment (Oulala-Leon) [reply app/src/main/java/sirup/edubook_api_test/MainActivity.java]: Set up a RecyclerView that works. However images are of an unrealistic size (Android console seems to be downloading dozens of 5MB files), so I'm now looking for a way to downsize them.

File: app/src/main/java/sirup/edubook_api_test/myHttpRequest.java [hunk 10]
```diff
@@ -1,72 +1,88 @@
...
-                            ret = new JSONObject(response);
-                        }
-                        catch (JSONException e) {
-                            e.printStackTrace();
-                        }
+                    public void onResponse(JSONArray response) {
+                        Log.d(TAG, "response :" + response);
+                        final JSONAdapter adapter = new JSONAdapter(response, _context);
+                        view.setAdapter(adapter);
+                        adapter.notifyDataSetChanged();
```
[the remark this answers is not in the document]
Comment (Oulala-Leon) [reply app/src/main/java/sirup/edubook_api_test/myHttpRequest.java]: Fixed.

**Pull request merged**